prometheus/internalmetrics/internal_metrics.go

```diff
@@ -0,0 +1,202 @@
+// Package internalmetrics provides functionality to collect Prometheus metrics.
+package internalmetrics
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	// Because this package is used all throughout the codebase, be VERY careful adding new
+	// kiali imports here. Most likely you will encounter an import cycle error that will
+	// cause a compilation failure.
+)
+
+// These constants define the different label names for the different metric timeseries
+const (
+	labelGraphKind        = "graph_kind"
+	labelGraphType        = "graph_type"
+	labelWithServiceNodes = "with_service_nodes"
+	labelAppender         = "appender"
+	labelName             = "name"
+	labelQueryID          = "query_id"
+)
+
+// MetricsType defines all of Kiali's own internal metrics.
+type MetricsType struct {
+	GraphsGenerated          *prometheus.CounterVec
+	GraphNodes               *prometheus.GaugeVec
+	GraphGenerationTime      *prometheus.HistogramVec
+	GraphAppenderTime        *prometheus.HistogramVec
+	GraphMarshalTime         *prometheus.HistogramVec
+	APIProcessingTime        *prometheus.HistogramVec
+	PrometheusProcessingTime *prometheus.HistogramVec
+}
+
+// Metrics contains all of Kiali's own internal metrics.
+// These metrics can be accessed directly to update their values, or
+// you can use available utility functions defined below.
+var Metrics = MetricsType{
+	GraphsGenerated: prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "kiali_graphs_generated_total",
+			Help: "The total number of graphs Kiali has generated.",
+		},
+		[]string{labelGraphKind, labelGraphType, labelWithServiceNodes},
+	),
+	GraphNodes: prometheus.NewGaugeVec(
```

Comment:
It doesn't feel right to have a gauge here while other metrics are counters.

Reply:
can't be a counter - graph nodes can go up or down depending on the graphs being generated and when deployments are added / removed. Gauge is appropriate here.